packages/api/portfolioDrafts/submit/submitPortfolioDraft.ts

```diff
@@ -0,0 +1,79 @@
+import { DynamoDBClient } from "@aws-sdk/client-dynamodb";
+import { DynamoDBDocumentClient, UpdateCommand, UpdateCommandOutput } from "@aws-sdk/lib-dynamodb";
+import { APIGatewayProxyEvent, APIGatewayProxyResult } from "aws-lambda";
+import { ErrorCodes } from "../../models/Error";
+import { ApiSuccessResponse, ErrorResponse, ErrorStatusCode, SuccessStatusCode } from "../../utils/response";
+import { isBodyPresent, isPathParameterPresent, isPortfolioStep, isValidJson } from "../../utils/validation";
+import { DATABASE_ERROR } from "../../utils/errors";
+import { PortfolioDraft } from "../../models/PortfolioDraft";
+import { v4 as uuidv4 } from "uuid";
+
+const TABLE_NAME = process.env.ATAT_TABLE_NAME ?? "";
+export const NO_SUCH_PORTFOLIO = new ErrorResponse(
+  { code: ErrorCodes.INVALID_INPUT, message: "Portfolio Draft with the given ID does not exist" },
+  ErrorStatusCode.NOT_FOUND
+);
+export const EMPTY_REQUEST_BODY = new ErrorResponse(
+  { code: ErrorCodes.INVALID_INPUT, message: "Request body must be empty" },
+  ErrorStatusCode.BAD_REQUEST
+);
+
+/**
+ * Submits all progress from the Portfolio Provisioning Wizard
+ *
+ * @param event - The POST request from API Gateway
+ */
+export async function handler(event: APIGatewayProxyEvent): Promise<APIGatewayProxyResult> {
+  if (isBodyPresent(event.body)) {
+    return EMPTY_REQUEST_BODY;
+  }
+
+  const portfolioDraftId = event.pathParameters?.portfolioDraftId;
+
+  if (!isPathParameterPresent(portfolioDraftId)) {
+    return NO_SUCH_PORTFOLIO;
+  }
+
+  try {
+    const result = await submitPortfolioDraftCommand(TABLE_NAME, portfolioDraftId);
+
+    return new ApiSuccessResponse(result.Attributes as PortfolioDraft, SuccessStatusCode.ACCEPTED);
+  } catch (error) {
+    if (error.name === "ConditionalCheckFailedException") {
+      return NO_SUCH_PORTFOLIO;
+    }
+    console.log("Database error: " + error.name);
+    return new ErrorResponse(
+      { code: ErrorCodes.OTHER, message: "Database error: " + error.name },
+      ErrorStatusCode.INTERNAL_SERVER_ERROR
+    );
+  }
+}
+export async function submitPortfolioDraftCommand(
+  table: string,
+  portfolioDraftId: string
+): Promise<UpdateCommandOutput> {
+  const dynamodb = new DynamoDBClient({});
+  const ddb = DynamoDBDocumentClient.from(dynamodb);
+  const now = new Date().toISOString();
+  const submitId = uuidv4();
+  const result = await ddb.send(
+    new UpdateCommand({
+      TableName: table,
+      Key: {
+        id: portfolioDraftId,
+      },
+      UpdateExpression: "set #portfolioVariable = :portfolio, updated_at = :now",
+      ExpressionAttributeNames: {
+        "#portfolioVariable": "submit_id",
+      },
```

Comment:
Confused about why one of these gets a name (`#portfolioVariable`->`submit_id`) but the other (`updated_at`) does not. Can we be consistent, Also `#portfolioVariable` is not a particularly descriptive name. If we were to make both of these variables, `#portfolioVariable` would be equally valid for both.

Reply:
That is because submit_id is a **new attribute** added to the object, for example:
```
{
  "created_at": "2021-09-07T20:34:13.647Z",
  "id": "c54d1d10-faab-4870-9e7c-9d33dfa6743e",
  "num_portfolio_managers": 0,
  "status": "not_started",
  "submit_id": "f22872c2-ef76-4be4-a124-54e366b8bbf4",
  "updated_at": "2021-09-07T20:55:30.337Z"
}
```
`updated_at` already exists on the object, so we don't have to set a new expression attribute for it.

As for why its named `#portfolioVariable` : [dynamodb docs: An expression attribute name must begin with a pound sign (#)](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Expressions.ExpressionAttributeNames.html)

We are doing the same thing for the createPortfolioStep, and createFundingStep:
```
 UpdateExpression: "set #portfolioVariable = :portfolio, updated_at = :now",
    ExpressionAttributeNames: {
      "#portfolioVariable": FUNDING_STEP,
    },
    ExpressionAttributeValues: {
      ":portfolio": fundingStep,
      ":now": now,
    },
```
`#portfolioVariable` is replaced with `FUNDING_STEP` in this instance, and `:portfolio` is set to the fundingStep object.
In the dynamodb docs, they usually use `#x = :y` for these replaced variables.

If we wanted to be cleaner, I guess we could do `#submitPortfolioVariable` instead.

Reply:
Hmm well I still don't like the name `#portfolioVariable`. Can we pick a better name here? I'd like `#submitId` honestly. Or `#newAttribute`. Or something like that.

Reply:
It's definitely confusing and it's hard to see why DynamoDB would require you to use a variable name for a variable itself regardless (seems like more of a fringe edge case when you want a dynamically generated attribute name). But if that's indeed required then so be it.

Reply:
Why add `submit_id` and not say, a datetime `submitted_at`?  Is it because something is needed to accompany `updated_at` which will have that datetime?  Does the presence of `submit_id` act as a flag to indicate this is no longer a Portfolio _Draft_?

Reply:
I wrestled with some of the issues discussed above in `feature/AT-6409`.  Here was my attempt at a consistent approach that could be used across steps.  The value of `stepKey` can even reference [one of these](https://github.com/dod-ccpo/atat-web-api/blob/d5cb5256f228c36ec4f9c57ae28d24906306ee60/packages/api/models/PortfolioDraft.ts#L7).
```
      new UpdateCommand({
        TableName: process.env.ATAT_TABLE_NAME ?? "",
        Key: {
          id: portfolioDraftId,
        },
        UpdateExpression: "set #stepKey = :step, #updateAtKey = :now",
        ExpressionAttributeNames: {
          // values are JSON keys
          "#stepKey": "funding_step",
          "#updateAtKey": "updated_at",
        },
        ExpressionAttributeValues: {
          ":step": fundingStep,
          ":now": new Date().toISOString(),
        },
        ConditionExpression: "attribute_exists(created_at)",
        ReturnValues: "ALL_NEW",
      })
```

Reply:
```
export async function submitPortfolioDraftCommand(
  table: string,
  portfolioDraftId: string
): Promise<UpdateCommandOutput> {
  const dynamodb = new DynamoDBClient({});
  const ddb = DynamoDBDocumentClient.from(dynamodb);
  const now = new Date().toISOString();
  const submitId = uuidv4();
  const result = await ddb.send(
    new UpdateCommand({
      TableName: table,
      Key: {
        id: portfolioDraftId,
      },
      UpdateExpression: "set #submitId = :submitIdValue, updated_at = :now",
      ExpressionAttributeNames: {
        "#submitId": "submit_id",
      },
      ExpressionAttributeValues: {
        ":now": now,
        ":submitIdValue": submitId,
      },
      ConditionExpression: "attribute_exists(created_at)",
      ReturnValues: "ALL_NEW",
    })
  );
  return result;
}
```
Is this better?

Reply:
@jasonburkert , the `submit_id` field is a placeholder that we'll leave in for now but is subject to change. We may end up nixing it given how we're going to track provisioning status, but that's something to deal with in M3.

Reply:
@jeffsegal-ccpo Thanks for clarification.

Reply:
@zachclark-ccpo That looks pretty good to me.  Only changes I would make:

1. eliminate `const submitId` and instead `":submitIdValue": uuidv4(),`
2. move table value into the Update Command creation unless table is needed elsewhere
`TableName: process.env.ATAT_TABLE_NAME ?? "",`

General question:  In the future would we want to log this as an important INFO-level event _"Jane Manager submitted Portfolio Draft ABC, submitId: XYZ"_